source/IAccessibleHandler/internalWinEventHandler.py

```diff
@@ -162,18 +181,33 @@ def _shouldGetEvents():
 	if _deferUntilForegroundWindow:
 		# #3831: Sometimes, a foreground event is fired,
 		# but GetForegroundWindow() takes a short while to return this new foreground.
+		curForegroundWindow = winUser.getForegroundWindow()
+		curForegroundClassName = winUser.getClassName(curForegroundWindow)
+		futureForegroundClassName = winUser.getClassName(_deferUntilForegroundWindow)
 		if (
 			_foregroundDefers < MAX_FOREGROUND_DEFERS
-			and winUser.getForegroundWindow() != _deferUntilForegroundWindow
+			and curForegroundWindow != _deferUntilForegroundWindow
 		):
 			# Wait a core cycle before handling events to give the foreground window time to update.
 			core.requestPump()
 			_foregroundDefers += 1
+			if isMSAADebugLoggingEnabled():
+				log.debugWarning(
+					f"Foreground still {curForegroundWindow} ({curForegroundClassName}). "
+					f"Deferring until foreground is {_deferUntilForegroundWindow} ({futureForegroundClassName}), "
+					f"defer count {_foregroundDefers}"
+				)
 			return False
 		else:
 			# Either the foreground window is now correct
 			# or we've already had the maximum number of defers.
 			# (Sometimes, foreground events are fired even when the foreground hasn't actually changed.)
+			if curForegroundWindow != _deferUntilForegroundWindow:
+				log.debugWarning(
+					"Foreground took too long to change. "
+					f"Foreground still {curForegroundWindow} ({curForegroundClassName}). "
+					f"Should be {_deferUntilForegroundWindow} ({futureForegroundClassName})"
+				)

```

Comment:
Note, this isn't protected by `isMSAADebugLoggingEnabled`, however I think that is appropriate for this message.